server/socketManager.js

```diff
@@ -0,0 +1,333 @@
+      }
+
+      socket.join('auth');
+      this.logger.info('Successful auth from %s.', socket.host);
+      this.handleAuth(socket);
+      return null;
+    });
+  }
+
+  /*
+   * Utility to verify that the pathname
+   * matches with the id of a client on the class
+   * @private
+   * @param {string} pathname - pathname to validate
+   * @returns {string} id
+   */
+  getIdFromPath(pathname) {
+    assert(
+      pathname[0] === '/' && pathname[pathname.length - 1] === '/',
+      'Invalid pathname'
+    );
+    const id = pathname.slice(1, pathname.length - 1);
+    if (!this.clients.has(id))
+      this.logger.warning(
+        `getIdFromPath: id ${id} from pathname ${pathname} does not exist`
+      );
+
+    return id;
+  }
+
+  /*
+   * Handle new auth'd websocket.
+   * @private
+   * @param {WebSocket} socket
+   */
+  handleAuth(socket) {
+    this.logger.info(`Preparing socket from ${socket.url}`);
+    const parseEvent = event => {
+      assert(typeof event === 'string', 'Must pass a string to getClient');
+      // defaulting to node client if there is no client prefix
+      let parsedEvent = event;
+      let id = 'node';
+
+      // check if first full word is a valid type
+      const split = event.indexOf(' ');
+      const prefix = event.slice(0, split);
+      if (this.types.indexOf(prefix) !== -1) {
+        id = prefix;
+        parsedEvent = event.slice(split + 1);
+      }
+      return [id, parsedEvent];
+    };
+
+    // pathname will be used similar to socket.io's namespaces
+    // they should correspond to the client id that will be used
+    // to make the calls
+    const { pathname } = new URL(socket.ws.url);
+    const id = this.getIdFromPath(pathname);
+    // broadcasts only send messages to the  node
+    // but originating socket does not expect a response
+    /**
+      // Example broadcast from client:
+      socket.fire('broadcast', 'set filter', '00000000000000000000');
+      // which will result in the following fire to bcoin server
+      nodeClient.socket.fire('set filter', '00000000000000000000');
+    **/
+    socket.bind('broadcast', (_event, ...args) => {
+      assert(
+        this.clients.has(id),
+        `No client ${id} for request from ${socket.url}`
+      );
+      const [clientId, event] = parseEvent(_event);
+      const client = this.clients.get(id)[clientId];
+      this.logger.info(`broadcast "${event}"" to ${id}'s ${clientId} client`);
+      client.call(event, ...args);
+    });
+
+    // requests from client to subscribe to events from node
+    // client should indicate the event to listen for
+    // and the `responseEvent` to fire when the event is heard
+    socket.bind('subscribe', async (_event, responseEvent) => {
+      assert(
+        this.clients.has(id),
+        `No client ${id} for request from ${socket.url}`
+      );
+
+      const [clientId, event] = parseEvent(_event);
+      const client = this.clients.get(id)[clientId];
```

Comment:
Feels unsafe and prone to an `undefined does not...` type of bug

```javascript
const clients = this.clients.get(id);
assert(clients)
const client = clients[cliendId];
```

Reply:
So this one is tough because of the way sockets work, kind of asynchronously. I was originally going to do something like what you're talking about (actually handle it in parseEvent), but something I realized is that what if at the time that a socket is being attached the client doesn't actually exist yet but gets added later? This should still work since it's just attached to the event and this is a pointer, so when the client gets attached this will then start working. So yeah, not sure if we want to block this.

Reply:
Can `this.clients.get(id)` return `null` or `undefined`? I don't think we want to block it either, but I do think we want to have a useful error message of sorts when `this.clients` doesn't have the `id`

Reply:
There is an assert at the beginning of the function to confirm that there is a client by that id, so `this.clients.get(id)` should not ever return `null` or `undefined`. I'll add an `if` with log statement for a client that doesn't exist though.